File: core/forensics.py
from flask import Blueprint, request
from core.utils import success, error, safe_decode
import os
import io
import base64
import hashlib
import zipfile
import struct
import math

forensics_bp = Blueprint('forensics', __name__)

UPLOAD_FOLDER = 'uploads'


# -- FILE UPLOAD HELPER --------------------------------------------------------
def get_uploaded_file():
    if 'file' not in request.files:
        return None, error('No file uploaded')
    f = request.files['file']
    if f.filename == '':
        return None, error('No file selected')
    data = f.read()
    return data, None


# -- FILE ANALYSIS -------------------------------------------------------------
@forensics_bp.route('/analyze', methods=['POST'])
def analyze_file():
    """Full file analysis - magic bytes, entropy, strings, hashes."""
    data, err = get_uploaded_file()
    if err:
        return err

    # Hashes
    hashes = {
        'md5':    hashlib.md5(data).hexdigest(),
        'sha1':   hashlib.sha1(data).hexdigest(),
        'sha256': hashlib.sha256(data).hexdigest(),
        'sha512': hashlib.sha512(data).hexdigest(),
    }

    # Entropy
    freq    = [0] * 256
    for b in data: freq[b] += 1
    n       = len(data)
    if n == 0:
        return error('Uploaded file is empty')
    entropy = 0.0
    for f in freq:
        if f > 0:
            p        = f / n
            entropy -= p * math.log2(p)

    # Magic bytes detection
    file_type = detect_magic(data)

    # Extract strings
    strings = extract_strings(data, min_len=4)

    # Hex preview (first 256 bytes)
    hex_preview = format_hex_dump(data[:256])

    return success({
        'filename':    request.files['file'].filename,
        'size':        len(data),
        'hashes':      hashes,
        'entropy':     round(entropy, 4),
        'file_type':   file_type,
        'strings':     strings[:200],  # limit
        'string_count':len(strings),
        'hex_preview': hex_preview,
        'is_text':     entropy < 5.0,
        'is_encrypted':entropy > 7.5,
    })


# -- STRINGS EXTRACTOR ---------------------------------------------------------
@forensics_bp.route('/strings', methods=['POST'])
def extract_strings_route():
    """Extract all printable strings from binary file."""
    data, err = get_uploaded_file()
    if err:
        return err

    min_len   = int(request.form.get('min_len', 4))
    filter_kw = request.form.get('filter', '').lower()
    strings   = extract_strings(data, min_len=min_len)

    if filter_kw:
        strings = [s for s in strings if filter_kw in s.lower()]

    return success({
        'total':   len(strings),
        'strings': strings[:2000],
        'filter':  filter_kw,
    })


def extract_strings(data, min_len=4):
    strings  = []
    current  = ''
    for b in data:
        if 0x20 <= b < 0x7f:
            current += chr(b)
        else:
            if len(current) >= min_len:
                strings.append(current)
            current = ''
    if len(current) >= min_len:
        strings.append(current)
    return strings


# -- HEX DUMP ------------------------------------------------------------------
@forensics_bp.route('/hexdump', methods=['POST'])
def hex_dump_route():
    """Generate full hex dump of uploaded file."""
    data, err = get_uploaded_file()
    if err:
        return err

    max_bytes = int(request.form.get('max_bytes', 4096))
    chunk     = data if max_bytes <= 0 else data[:max_bytes]
    dump      = format_hex_dump(chunk)

    return success({
        'total_size':   len(data),
        'shown_bytes':  len(chunk),
        'hex_dump':     dump,
        'truncated':    max_bytes > 0 and len(data) > max_bytes,
    })


def format_hex_dump(data):
    lines = []
    for i in range(0, len(data), 16):
        chunk   = data[i:i+16]
        offset  = f'{i:08x}'
        hex_str = ' '.join(f'{b:02x}' for b in chunk).ljust(47)
        ascii_  = ''.join(chr(b) if 0x20 <= b < 0x7f else '.' for b in chunk)
        lines.append(f'{offset}  {hex_str}  |{ascii_}|')
    return '\n'.join(lines)


# -- MAGIC BYTES DETECTOR ------------------------------------------------------
@forensics_bp.route('/magic', methods=['POST'])
def magic_bytes_route():
    """Identify file type from magic bytes."""
    data, err = get_uploaded_file()
    if err:
        return err

    result = detect_magic(data)
    hex_bytes = ' '.join(f'{b:02x}' for b in data[:16])

    return success({
        'file_type':  result,
        'first_bytes':hex_bytes,
        'filename':   request.files['file'].filename,
    })


def detect_magic(data):
    if len(data) < 4:
        return 'Unknown (too small)'

    signatures = [
        (b'\x89PNG\r\n\x1a\n',          'PNG Image'),
        (b'\xff\xd8\xff',                'JPEG Image'),
        (b'GIF8',                         'GIF Image'),
        (b'BM',                           'BMP Image'),
        (b'%PDF',                         'PDF Document'),
        (b'PK\x03\x04',                   'ZIP Archive'),
        (b'Rar!\x1a\x07',                 'RAR Archive'),
        (b'\x1f\x8b\x08',                 'GZIP Archive'),
        (b'BZh',                           'BZIP2 Archive'),
        (b'\xfd7zXZ\x00',                  'XZ Archive'),
        (b'\x7fELF',                       'ELF Binary'),
        (b'MZ',                            'Windows PE/EXE'),
        (b'\xca\xfe\xba\xbe',              'Java Class File'),
        (b'\xce\xfa\xed\xfe',              'Mach-O 32-bit'),
        (b'\xcf\xfa\xed\xfe',              'Mach-O 64-bit'),
        (b'RIFF',                           'WAV/AVI'),
        (b'\x00\x00\x00\x18ftyp',           'MP4 Video'),
        (b'SQLite format 3\x00',            'SQLite Database'),
        (b'#!',                             'Script/Shebang'),
        (b'OggS',                           'OGG Audio'),
        (b'ID3',                            'MP3 Audio'),
        (b'\x49\x49\x2a\x00',              'TIFF (Little Endian)'),
        (b'\x4d\x4d\x00\x2a',              'TIFF (Big Endian)'),
        (b'IHDR',                           'PNG Chunk'),
        (b'-----BEGIN',                     'PEM Certificate/Key'),
        (b'<?xml',                          'XML Document'),
        (b'<html',                          'HTML Document'),
        (b'{',                              'JSON (possible)'),
    ]

    for sig, name in signatures:
        if data[:len(sig)] == sig:
            return name

    # Check if text
    try:
        data[:512].decode('utf-8')
        return 'Text/ASCII'
    except Exception:
        pass

    return 'Unknown Binary'


# -- PNG CHUNK ANALYZER --------------------------------------------------------
@forensics_bp.route('/png_chunks', methods=['POST'])
def png_chunks():
    """Analyze PNG file chunks."""
    data, err = get_uploaded_file()
    if err:
        return err

    if data[:8] != b'\x89PNG\r\n\x1a\n':
        return error('Not a valid PNG file')

    chunks  = []
    offset  = 8
    while offset < len(data) - 12:
        try:
            length = struct.unpack('>I', data[offset:offset+4])[0]
            ctype  = data[offset+4:offset+8].decode('ascii', errors='replace')
            cdata  = data[offset+8:offset+8+length]
            crc    = struct.unpack('>I', data[offset+8+length:offset+12+length])[0]

            preview = ''
            if ctype in ('tEXt', 'iTXt', 'zTXt'):
                try:
                    preview = cdata.decode('utf-8', errors='replace').replace('\x00', ' | ')
                except Exception:
                    pass

            chunks.append({
                'type':    ctype,
                'length':  length,
                'offset':  offset,
                'crc':     hex(crc),
                'preview': preview[:200],
                'hex':     cdata[:16].hex() + ('...' if len(cdata) > 16 else ''),
            })
            offset += 12 + length
            if ctype == 'IEND':
                break
        except Exception:
            break

    return success({
        'valid': True,
        'chunks': chunks,
        'total':  len(chunks),
    })


# -- ENTROPY ANALYSIS ----------------------------------------------------------
@forensics_bp.route('/entropy', methods=['POST'])
def entropy_analysis():
    """Calculate Shannon entropy and byte frequency."""
    data, err = get_uploaded_file()
    if err:
        return err

    freq    = [0] * 256
    for b in data: freq[b] += 1
    n       = len(data)
    if n == 0:
        return error('Uploaded file is empty')
    entropy = 0.0
    for f in freq:
        if f > 0:
            p        = f / n
            entropy -= p * math.log2(p)

    # Top bytes
    top_bytes = sorted(
        [{ 'byte': i, 'hex': hex(i), 'count': freq[i], 'pct': round(freq[i]/n*100, 2) }
         for i in range(256) if freq[i] > 0],
        key=lambda x: x['count'], reverse=True
    )[:20]

    interpretation = (
        'Very high - likely encrypted or compressed' if entropy > 7.5
        else 'High - possibly compressed or binary' if entropy > 6.5
        else 'Medium - mixed data' if entropy > 4.5
        else 'Low - likely plain text or structured data'
    )

    return success({
        'entropy':        round(entropy, 6),
        'max_entropy':    8.0,
        'interpretation': interpretation,
        'file_size':      n,
        'unique_bytes':   sum(1 for f in freq if f > 0),
        'null_bytes':     freq[0],
        'top_bytes':      top_bytes,
    })


# -- ZIP PASSWORD CRACKER ------------------------------------------------------
@forensics_bp.route('/zip_crack', methods=['POST'])
def zip_crack():
    """Crack ZIP file password using provided wordlist."""
    if 'file' not in request.files:
        return error('No ZIP file uploaded')

    wordlist_raw = request.form.get('wordlist', '')
    if not wordlist_raw:
        return error('Provide a wordlist (newline separated)')

    zip_data = request.files['file'].read()
    words    = [w.strip() for w in wordlist_raw.split('\n') if w.strip()]

    if len(words) > 50000:
        return error('Wordlist too large (max 50,000 words)')

    try:
        import pyzipper
        zf = pyzipper.AESZipFile(io.BytesIO(zip_data))

        names = zf.namelist()
        if not names:
            return error('ZIP archive is empty')

        # Test passwords by reading the first file instead of extracting files.
        try:
            zf.read(names[0], pwd=b'')
            return success({ 'protected': False, 'message': 'ZIP is not password protected' })
        except Exception:
            pass

        for word in words:
            try:
                zf.read(names[0], pwd=word.encode('utf-8'))
                return success({
                    'found':    True,
                    'password': word,
                    'tried':    words.index(word) + 1,
                })
            except Exception:
                pass

        return success({
            'found':   False,
            'tried':   len(words),
            'message': 'Password not found in wordlist',
        })

    except Exception as ex:
        # Fallback to standard zipfile
        try:
            zf = zipfile.ZipFile(io.BytesIO(zip_data))
            names = zf.namelist()
            if not names:
                return error('ZIP archive is empty')
            for word in words:
                try:
                    zf.read(names[0], pwd=word.encode())
                    return success({ 'found': True, 'password': word })
                except Exception:
                    pass
            return success({ 'found': False, 'tried': len(words) })
        except Exception as ex2:
            return error(f'ZIP crack error: {str(ex2)}')


# -- LSB STEGANOGRAPHY ---------------------------------------------------------
@forensics_bp.route('/lsb', methods=['POST'])
def lsb_extract():
    """Extract LSB steganography from image (server-side with Pillow)."""
    data, err = get_uploaded_file()
    if err:
        return err

    bits_per_channel = int(request.form.get('bits', 1))

    try:
        from PIL import Image
        img    = Image.open(io.BytesIO(data)).convert('RGB')
        pixels = list(img.getdata())

        mask   = (1 << bits_per_channel) - 1
        bits   = ''

        for r, g, b in pixels:
            bits += format(r & mask, f'0{bits_per_channel}b')
            bits += format(g & mask, f'0{bits_per_channel}b')
            bits += format(b & mask, f'0{bits_per_channel}b')

        # Convert to bytes
        byte_arr = []
        for i in range(0, len(bits) - 8, 8):
            byte_arr.append(int(bits[i:i+8], 2))

        # Find null terminator
        text = ''
        for b in byte_arr:
            if b == 0:
                break
            if 0x20 <= b < 0x7f:
                text += chr(b)
            elif text:
                text += '.'

        hex_preview = ' '.join(f'{b:02x}' for b in byte_arr[:32])

        has_flag   = any(p in text.lower() for p in ['flag{', 'ctf{', 'htb{'])
        printable  = sum(1 for b in byte_arr[:1000] if 0x20 <= b < 0x7f)
        sample_size = min(len(byte_arr), 1000) or 1
        p_ratio    = printable / sample_size

        return success({
            'image_size':      f'{img.width}x{img.height}',
            'bits_per_channel': bits_per_channel,
            'extracted_text':  text[:500],
            'hex_preview':     hex_preview,
            'printable_ratio': round(p_ratio, 3),
            'has_flag':        has_flag,
            'total_bytes':     len(byte_arr),
        })

    except ImportError:
        return error('Pillow not installed. Run: pip install Pillow')
    except Exception as ex:
        return error(f'LSB extraction failed: {str(ex)}')

File: static/js/tools/forensics.js
function buildFileAnalysis(panel){
    panel.innerHTML = `
    ${toolHeader('','Full File Analysis','Magic bytes, hashes, entropy, strings - all in one shot')}
    <div class="tool-wrap">
        <div class="tool-title"> Full File Analysis</div>
        <label>Upload File</label>
        <input type="file" id="analysisFile">
        <label>Flag format (optional)</label>
        <input type="text" id="analysisFlagFormat" placeholder="flag{} or picoCTF{...}">
        <div class="button-group">
            <button class="btn btn-run" onclick="runFileAnalysis()">Analyze</button>
            <button class="btn btn-outline" onclick="clearFileAnalysis()">Clear</button>
        </div>
        ${createOutput('analysisOutput','Analysis Results')}
    </div>`;
}

async function runFileAnalysis(){
    const file = document.getElementById('analysisFile').files[0];
    const flagFormat = document.getElementById('analysisFlagFormat').value.trim();
    if(!file){ showToast('Upload a file','error'); return; }
    setLoading('analysisOutput');
    const res = await apiUpload('/api/forensics/analyze', file);
    if(res.status === 'ok'){
        const d = res.data;
        let out = `File: ${d.filename}\nSize: ${d.size.toLocaleString()} bytes\n`;
        out += `Type: ${d.file_type}\n`;
        out += `Entropy: ${d.entropy} bits/byte\n`;
        out += `Text file: ${d.is_text}\nEncrypted/Compressed: ${d.is_encrypted}\n\n`;
        out += `HASHES:\n  MD5:    ${d.hashes.md5}\n  SHA1:   ${d.hashes.sha1}\n  SHA256: ${d.hashes.sha256}\n\n`;
        out += `HEX PREVIEW (first 256 bytes):\n${d.hex_preview}\n\n`;
        out += `STRINGS (first 50):\n${d.strings.slice(0,50).map(s=>'  '+s).join('\n')}`;
        if(d.string_count > 50) out += `\n  ... and ${d.string_count-50} more strings`;
        if(flagFormat){
            const stringsText = d.strings.join('\n');
            const flagRes = await findFlags(stringsText, flagFormat);
            if(flagRes.status === 'ok'){
                setOutput('analysisOutput',
                    `<div class="plain-output-block">${escapeHtml(out)}</div>` +
                    flagResultsHtml(stringsText, flagRes.data, {
                        title: 'Flags found in strings',
                        sourceLabel: 'Extracted Strings'
                    }),
                    true
                );
            } else {
                setOutput('analysisOutput', `${out}\n\nFlag search error: ${flagRes.error}`);
            }
        } else {
            setOutput('analysisOutput', out);
        }
        showToast('Analysis complete!');
    } else {
        setOutput('analysisOutput', `Error: ${res.error}`);
    }
}
function clearFileAnalysis(){
    document.getElementById('analysisFile').value='';
    document.getElementById('analysisFlagFormat').value='';
    resetOutput('analysisOutput');
}

function buildStringsExtract(panel){
    panel.innerHTML = `
    ${toolHeader('','Strings Extractor','Extract printable strings with filter support')}
    <div class="tool-wrap">
        <div class="tool-title"> Strings Extractor</div>
        <label>Upload File</label>
        <input type="file" id="stringsFile">
        <div style="display:grid;grid-template-columns:1fr 1fr;gap:12px;">
            <div><label>Min Length</label><input type="number" id="stringsMin" value="4" min="1"></div>
            <div><label>Filter keyword (optional)</label><input type="text" id="stringsFilter" placeholder="flag, pass, key..."></div>
        </div>
        <label>Flag format (optional)</label>
        <input type="text" id="stringsFlagFormat" placeholder="flag{} or picoCTF{...}">
        <div class="button-group">
            <button class="btn btn-run" onclick="runStringsExtract()">Extract</button>
            <button class="btn btn-outline" onclick="clearStrings()">Clear</button>
        </div>
        ${createOutput('stringsOutput','Strings')}
    </div>`;
}

async function runStringsExtract(){
    const file   = document.getElementById('stringsFile').files[0];
    const minLen = document.getElementById('stringsMin').value;
    const filter = document.getElementById('stringsFilter').value;
    const flagFormat = document.getElementById('stringsFlagFormat').value.trim();
    if(!file){ showToast('Upload a file','error'); return; }
    setLoading('stringsOutput');
    const fd = new FormData();
    fd.append('file', file);
    fd.append('min_len', minLen);
    fd.append('filter', filter);
    const res = await apiUploadForm('/api/forensics/strings', fd);
    if(res.status === 'ok'){
        const d = res.data;
        const header = `Total strings: ${d.total}${d.filter ? ` (filtered by "${d.filter}")` : ''}\n\n`;
        const stringsText = d.strings.join('\n');
        if(flagFormat){
            const flagRes = await findFlags(stringsText, flagFormat);
            if(flagRes.status === 'ok'){
                setOutput('stringsOutput',
                    `<div class="plain-output-block">${escapeHtml(header)}</div>` +
                    flagResultsHtml(stringsText, flagRes.data, {
                        title: 'Flags found',
                        sourceLabel: 'Strings'
                    }),
                    true
                );
            } else {
                setOutput('stringsOutput', `${header}${stringsText}\n\nFlag search error: ${flagRes.error}`);
            }
        } else {
            setOutput('stringsOutput', header + stringsText);
        }
    } else {
        setOutput('stringsOutput', `Error: ${res.error}`);
    }
}
function clearStrings(){
    document.getElementById('stringsFile').value='';
    document.getElementById('stringsFlagFormat').value='';
    resetOutput('stringsOutput');
}

function buildHexDumpTool(panel){
    panel.innerHTML = `
    ${toolHeader('','Hex Dump','Full hex dump with offset and ASCII representation')}
    <div class="tool-wrap">
        <div class="tool-title"> Hex Dump</div>
        <label>Upload File</label>
        <input type="file" id="hexDumpFile">
        <label>Max bytes to show</label>
        <select id="hexDumpMax">
            <option value="512">512</option>
            <option value="1024">1 KB</option>
            <option value="4096" selected>4 KB</option>
            <option value="16384">16 KB</option>
            <option value="0">All</option>
        </select>
        <div class="button-group">
            <button class="btn btn-run" onclick="runHexDumpTool()">Dump</button>
            <button class="btn btn-outline" onclick="clearHexDump()">Clear</button>
        </div>
        ${createOutput('hexDumpOutput','Hex Dump')}
    </div>`;
}

async function runHexDumpTool(){
    const file = document.getElementById('hexDumpFile').files[0];
    const max  = document.getElementById('hexDumpMax').value;
    if(!file){ showToast('Upload a file','error'); return; }
    setLoading('hexDumpOutput');
    const fd = new FormData();
    fd.append('file', file);
    fd.append('max_bytes', max);
    const res = await apiUploadForm('/api/forensics/hexdump', fd);
    if(res.status === 'ok'){
        const d = res.data;
        let header = `File size: ${d.total_size.toLocaleString()} bytes | Showing: ${d.shown_bytes.toLocaleString()} bytes\n\n`;
        setOutput('hexDumpOutput', header + d.hex_dump);
    } else {
        setOutput('hexDumpOutput', `Error: ${res.error}`);
    }
}
function clearHexDump(){ document.getElementById('hexDumpFile').value=''; resetOutput('hexDumpOutput'); }

function buildZipCrack(panel){
    panel.innerHTML = `
    ${toolHeader('','ZIP Password Cracker','Crack password-protected ZIP files using wordlist')}
    <div class="tool-wrap">
        <div class="tool-title"> ZIP Password Cracker</div>
        <label>Upload ZIP File</label>
        <input type="file" id="zipFile" accept=".zip">
        <label>Wordlist (one password per line)</label>
        <textarea id="zipWordlist" style="min-height:120px;"
            placeholder="password&#10;123456&#10;admin&#10;secret&#10;..."></textarea>
        <div class="button-group">
            <button class="btn btn-run" onclick="runZipCrack()">Crack ZIP</button>
            <button class="btn btn-outline" onclick="clearZipCrack()">Clear</button>
        </div>
        ${createOutput('zipOutput','Crack Result')}
    </div>`;
}

async function runZipCrack(){
    const file     = document.getElementById('zipFile').files[0];
    const wordlist = document.getElementById('zipWordlist').value.trim();
    if(!file){ showToast('Upload a ZIP file','error'); return; }
    if(!wordlist){ showToast('Enter a wordlist','error'); return; }
    setLoading('zipOutput');
    const fd = new FormData();
    fd.append('file', file);
    fd.append('wordlist', wordlist);
    const res = await apiUploadForm('/api/forensics/zip_crack', fd);
    if(res.status === 'ok'){
        const d = res.data;
        if(d.found){
            setOutput('zipOutput', `OK PASSWORD FOUND!\n\nPassword: ${d.password}\nTried: ${d.tried} passwords`);
            showToast(`Password found: ${d.password}`);
        } else if(d.protected === false){
            setOutput('zipOutput', 'ZIP is not password protected');
        } else {
            setOutput('zipOutput', `NO Not found\nTried: ${d.tried} passwords\n\n${d.message || ''}`);
        }
    } else {
        setOutput('zipOutput', `Error: ${res.error}`);
    }
}
function clearZipCrack(){
    document.getElementById('zipFile').value='';
    document.getElementById('zipWordlist').value='';
    resetOutput('zipOutput');
}

function buildLsbStego(panel){
    panel.innerHTML = `
    ${toolHeader('','LSB Steganography','Server-side LSB extraction using Pillow')}
    <div class="tool-wrap">
        <div class="tool-title"> LSB Stego Extractor</div>
        <label>Upload Image (PNG recommended)</label>
        <input type="file" id="lsbFile" accept="image/*">
        <label>Bits per channel</label>
        <select id="lsbBits">
            <option value="1">1 bit</option>
            <option value="2">2 bits</option>
            <option value="4">4 bits</option>
        </select>
        <label>Flag format (optional)</label>
        <input type="text" id="lsbFlagFormat" placeholder="flag{} or picoCTF{...}">
        <div class="button-group">
            <button class="btn btn-run" onclick="runLsbStego()">Extract LSB</button>
            <button class="btn btn-outline" onclick="clearLsb()">Clear</button>
        </div>
        ${createOutput('lsbOutput','Extracted Data')}
    </div>`;
}

async function runLsbStego(){
    const file = document.getElementById('lsbFile').files[0];
    const bits = document.getElementById('lsbBits').value;
    const flagFormat = document.getElementById('lsbFlagFormat').value.trim();
    if(!file){ showToast('Upload an image','error'); return; }
    setLoading('lsbOutput');
    const fd = new FormData();
    fd.append('file', file);
    fd.append('bits', bits);
    const res = await apiUploadForm('/api/forensics/lsb', fd);
    if(res.status === 'ok'){
        const d = res.data;
        let out = `Image: ${d.image_size} | Bits/channel: ${d.bits_per_channel}\n`;
        out += `Printable ratio: ${(d.printable_ratio*100).toFixed(1)}%\n`;
        out += `Flag detected: ${d.has_flag ? ' YES' : 'No'}\n\n`;
        out += `Hex preview: ${d.hex_preview}\n\n`;
        out += `ASCII text:\n${d.extracted_text || '(no printable text found)'}`;
        if(flagFormat){
            const text = d.extracted_text || '';
            const flagRes = await findFlags(text, flagFormat);
            if(flagRes.status === 'ok'){
                setOutput('lsbOutput',
                    `<div class="plain-output-block">${escapeHtml(out)}</div>` +
                    flagResultsHtml(text, flagRes.data, {
                        title: 'Flags found',
                        sourceLabel: 'Extracted Text'
                    }),
                    true
                );
            } else {
                setOutput('lsbOutput', `${out}\n\nFlag search error: ${flagRes.error}`);
            }
        } else {
            setOutput('lsbOutput', out);
        }
        if(d.has_flag) showToast(' Flag detected in LSB!');
    } else {
        setOutput('lsbOutput', `Error: ${res.error}`);
    }
}
function clearLsb(){
    document.getElementById('lsbFile').value='';
    document.getElementById('lsbFlagFormat').value='';
    resetOutput('lsbOutput');
}

function buildEntropyTool(panel){
    panel.innerHTML = `
    ${toolHeader('','Entropy Analysis','Shannon entropy and byte frequency analysis')}
    <div class="tool-wrap">
        <div class="tool-title"> Entropy Analyzer</div>
        <label>Upload File</label>
        <input type="file" id="entropyFile">
        <div class="button-group">
            <button class="btn btn-run" onclick="runEntropyTool()">Analyze</button>
            <button class="btn btn-outline" onclick="clearEntropy()">Clear</button>
        </div>
        ${createOutput('entropyOutput','Entropy Results')}
    </div>`;
}

async function runEntropyTool(){
    const file = document.getElementById('entropyFile').files[0];
    if(!file){ showToast('Upload a file','error'); return; }
    setLoading('entropyOutput');
    const res = await apiUpload('/api/forensics/entropy', file);
    if(res.status === 'ok'){
        const d = res.data;
        let out = `Entropy: ${d.entropy} / 8.0 bits/byte\n`;
        out += `Interpretation: ${d.interpretation}\n\n`;
        out += `File size: ${d.file_size.toLocaleString()} bytes\n`;
        out += `Unique bytes: ${d.unique_bytes} / 256\n`;
        out += `Null bytes: ${d.null_bytes}\n\n`;
        out += `TOP 20 BYTES:\n`;
        d.top_bytes.forEach(b => {
            out += `  0x${b.byte.toString(16).padStart(2,'0')} (${b.byte.toString().padStart(3)})  ${b.count.toLocaleString().padStart(8)}  ${b.pct}%\n`;
        });
        setOutput('entropyOutput', out);
    } else {
        setOutput('entropyOutput', `Error: ${res.error}`);
    }
}
function clearEntropy(){ document.getElementById('entropyFile').value=''; resetOutput('entropyOutput'); }
